### DISCUSSION TITLE
Puzzling behavior with constrained optimization

### DISCUSSION THREAD

**Discussion by vivien000** [google/jaxopt#252] [Q&A] [answered]
Hi. I'm trying to solve a constrained optimization problem but I face a puzzling behavior that I don't understand. Any help would be welcome and sorry if I made an obvious mistake (I'm a beginner with Jaxopt).

To simplify, let's say that I'd like to solve:
$$\min(x_1^2 + x_2^2 + x_3^2 + x_4^2)$$
with:
- $x_1 + x_2 = 100$
- $x_3 + x_4 = 100$
- $x_1 + x_3 = 90$
- $x_2 + x_4 = 110$

If I do the following (cf. this [Colab notebook](https://colab.research.google.com/drive/1LahAYHc60cOClvje_HvVehklYLkljjMP?usp=sharing) with all the following code snippets), I get the expected result (i.e. [45, 55, 45, 55]).
```
import numpy as np
import jax.numpy as jnp
from jaxopt import ProjectedGradient
from jaxopt.projection import projection_polyhedron, projection_affine_set

def objective_fun(v):
  return jnp.sum(jnp.power(v, 2))

w_init = np.array([50., 50., 40., 60.])

hyperparams_proj = (
    np.array([
      [1., 1., 0., 0.],
      [0., 0., 1., 1.],
      [1., 0., 1., 0.],
      [0., 1., 0., 1.],
    ]), 
    np.array([
      100.,
      100.,
      90.,
      110.
    ])
)

pg = ProjectedGradient(
    fun=objective_fun,
    projection=projection_affine_set
)
pg.run(w_init, hyperparams_proj=hyperparams_proj).params
```

# Problem 1

However, if I remove the fourth constraint (which a linear combination of the three other constraints and should then be superfluous), I get an incorrect result (`DeviceArray([nan, nan, nan, nan], dtype=float32)`).

```
hyperparams_proj = (
    np.array([
      [1., 1., 0., 0.],
      [0., 0., 1., 1.],
      [1., 0., 1., 0.]
    ]), 
    np.array([
      100.,
      100.,
      90.
    ])
)

pg = ProjectedGradient(
    fun=objective_fun,
    projection=projection_affine_set
)
pg.run(w_init, hyperparams_proj=hyperparams_proj).params
```

# Problem 2

Moreover, if I also want to enforce a positivity constraint on $x_1, x_2, x_3, x_4$ with `projection_polyhedron`, then I get a `ConcretizationTypeError`:

```
hyperparams_proj = (
    np.array([
      [1., 1., 0., 0.],
      [0., 0., 1., 1.],
      [1., 0., 1., 0.],
      [0., 1., 0., 1.],
    ]), 
    np.array([
      100.,
      100.,
      90.,
      110.
    ]),
      np.array([
      [-1., 0., 0., 0.],
      [0., -1., 0., 0.],
      [0., 0., -1., 0.],
      [0., 0., 0., -1.]
    ]), 
    np.array([0., 0., 0., 0.])
)

pg = ProjectedGradient(
    fun=objective_fun,
    projection=projection_polyhedron
)
pg.run(w_init, hyperparams_proj=hyperparams_proj).params
```

**Comment by Algue-Rythme** [comment, marked as answer]
Good catch. I see two unrelated problems here.

## Problem 1

Currently, `projection_affine_set` relies on `EqualityConstrainedQP` which is [known to be instable](https://jaxopt.github.io/stable/quadratic_programming.html#quadratic-programming) in some circumstances - specially in `float32` with large magnitude on `w_init` as it is the case here: the value of the objective is above `1e4` initially.

I see one easy hot fix : switch to `float64` by adding `jax.config.update("jax_enable_x64", True)` at the beginning of the program.

Ideally, one would want to control hyper-parameters `tol` and `refine_regularization=1e-3` [in call to EqualityConstrainedQP](https://github.com/google/jaxopt/blob/fdaf31d7e369c0bcdf46705c86d068d56f0d3d90/jaxopt/_src/projection.py#L278) solver within `projection_affine_set` : see [here](https://jaxopt.github.io/stable/_autosummary/jaxopt.EqualityConstrainedQP.html#jaxopt.EqualityConstrainedQP) 

But currently you need to re-implement `projection_affine_set` yourself to add those modifications.

I also noticed that setting your own step size by setting argument `stepsize=1e-2` in `ProjectedGradient` (without relying on [FISTA line search](https://jaxopt.github.io/stable/_autosummary/jaxopt.ProjectedGradient.html#jaxopt.ProjectedGradient.stepsize) which is the default behavior) seems to solve the issue.

## Problem 2

Initially the function `projection_polyhedron` was not jittable and code for jitting is lacking. Now [the underlying OSQP solver is jittable](https://github.com/google/jaxopt/blob/fdaf31d7e369c0bcdf46705c86d068d56f0d3d90/jaxopt/_src/projection.py#L303) - unfortunately [this line](https://github.com/google/jaxopt/blob/fdaf31d7e369c0bcdf46705c86d068d56f0d3d90/jaxopt/_src/projection.py#L329) is not.

The hot-fix would be to stop the feasability check as done below:

```python
def projection_polyhedron_no_feasability_check(x, hyperparams):
  return projection_polyhedron(x, hyperparams, check_feasible=False)

pg = ProjectedGradient(
    fun=objective_fun,
    projection=projection_polyhedron_no_feasability_check
)
pg.run(w_init, hyperparams_proj=hyperparams_proj).params
```

Once again, in a ideal situation we would like to pass those hyper-parameters directly to the projection method ourselves.

**Reply by vivien000** [reply]
Many thanks Louis
